connector/github/github.go

```diff
@@ -233,37 +243,82 @@ func (c *githubConnector) HandleCallback(s connector.Scopes, r *http.Request) (i
 	return identity, nil
 }
 
-func (c *githubConnector) Refresh(ctx context.Context, s connector.Scopes, ident connector.Identity) (connector.Identity, error) {
-	if len(ident.ConnectorData) == 0 {
-		return ident, errors.New("no upstream access token found")
+func (c *githubConnector) Refresh(ctx context.Context, s connector.Scopes, identity connector.Identity) (connector.Identity, error) {
+	if len(identity.ConnectorData) == 0 {
+		return identity, errors.New("no upstream access token found")
 	}
 
 	var data connectorData
-	if err := json.Unmarshal(ident.ConnectorData, &data); err != nil {
-		return ident, fmt.Errorf("github: unmarshal access token: %v", err)
+	if err := json.Unmarshal(identity.ConnectorData, &data); err != nil {
+		return identity, fmt.Errorf("github: unmarshal access token: %v", err)
 	}
 
 	client := c.oauth2Config(s).Client(ctx, &oauth2.Token{AccessToken: data.AccessToken})
 	user, err := c.user(ctx, client)
 	if err != nil {
-		return ident, fmt.Errorf("github: get user: %v", err)
+		return identity, fmt.Errorf("github: get user: %v", err)
 	}
 
 	username := user.Name
 	if username == "" {
 		username = user.Login
 	}
-	ident.Username = username
-	ident.Email = user.Email
+	identity.Username = username
+	identity.Email = user.Email
+
+	if s.Groups && len(c.orgs) > 0 {
+		groups, isAuthorized, err := c.isUserAuthorized(ctx, client, username)
+		if err != nil {
+			return identity, err
+		}
+		if !isAuthorized {
+			return identity, fmt.Errorf("github: user %q not a member of any orgs or teams provided", username)
+		}
+		// TODO(estroz): is a flat group slice of teams from potentially multiple orgs usable?
+		identity.Groups = groups
+	}
+
+	return identity, nil
+}
 
-	if s.Groups && c.org != "" {
-		groups, err := c.teams(ctx, client, c.org)
+// isUserAuthorized enforces org and team constraints on user authorization
+// Cases in which user is authorized:
+// 	N orgs, no teams: user is member of at least 1 org
+// 	N orgs, M teams per org: user is member of any team from at least 1 org
+// 	N-1 orgs, M teams per org, 1 org with no teams: user is member of any team from at least 1 org, or member of org with no teams
+func (c *githubConnector) isUserAuthorized(ctx context.Context, client *http.Client, userName string) (groups []string, isAuthorized bool, err error) {
+	var (
+		inOrg bool
+		teams []string
+	)
+	for _, org := range c.orgs {
+		teams, inOrg, err = c.userInOrgTeams(ctx, client, userName, org)
 		if err != nil {
-			return ident, fmt.Errorf("github: get teams: %v", err)
+			return
 		}
-		ident.Groups = groups
+		if !inOrg {
+			c.logger.Debugf("github: user %q not in org %q", userName, org.Name)
+			continue
+		}
+		if len(org.Teams) == 0 {
+			// User is in at least in one org with no team restrictions
+			isAuthorized = true
+			groups = append(groups, teams...)
+			c.logger.Debugf("github: user %q in org %q", userName, org.Name)
+			continue
```

Comment:
Really confused by this `continue` and the one following it. Not clear what you're trying to do here.